### PR TITLE
Pu/datefunctions

### PR DESCRIPTION
**Pull request by izulin**
### Context

15 datetime functions WEEKDAY, DATEVALUE, HOUR, MINUTE, SECOND, TIME, TIMEVALUE, NOW, TODAY, EDATE, WEEKNUM, ISOWEEKNUM, DATEDIF, DAYS360, YEARFRAC,
small refactor of datetime module
a lot of bugfixes in the existing datetime functions

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1.
2.
3.

### Checklist:

- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation,
- [x] I described the modification in the CHANGELOG.md file.

### CODE REVIEW

File: src/interpreter/plugin/DatePlugin.ts [hunk 1]
```diff
@@ -45,23 +78,96 @@ export class DatePlugin extends FunctionPlugin {
...
+          {argumentType: ArgumentTypes.NUMBER, minValue: 0},
         ]
     },
     'DAYS': {
       method: 'days',
-      parameters:[
-          {argumentType: ArgumentTypes.NUMBER},
-          {argumentType: ArgumentTypes.NUMBER},
+      parameters: [
+          {argumentType: ArgumentTypes.NUMBER, minValue: 0},
```
Comment (bardek8) [inline src/interpreter/plugin/DatePlugin.ts:94]: Why minValue: 0?

Comment (izulin) [reply src/interpreter/plugin/DatePlugin.ts:94]: XL compatibility

GS allows for negative values

File: src/interpreter/plugin/DatePlugin.ts [hunk 2]
```diff
@@ -45,23 +78,96 @@ export class DatePlugin extends FunctionPlugin {
...
+      method: 'edate',
+      parameters: [
+          {argumentType: ArgumentTypes.NUMBER, minValue: 0},
+          {argumentType: ArgumentTypes.NUMBER},
+        ],
+    },
+    'DATEDIF': {
+      method: 'datedif',
+      parameters: [
+        {argumentType: ArgumentTypes.NUMBER, minValue: 0},
```
[the remark this answers is not in the document]
Comment (izulin) [reply src/interpreter/plugin/DatePlugin.ts:150]: see reply above

File: src/interpreter/plugin/DatePlugin.ts [hunk 3]
```diff
@@ -153,4 +288,208 @@ export class DatePlugin extends FunctionPlugin {
...
+              this.interpreter.dateHelper.numberToSimpleDate(endDate)
+            )
+          case 2:
+            return (endDate-startDate)/360
+          case 3:
+            return (endDate-startDate)/365
+          case 4:
+            return this.days360Core(startDate, endDate, true) / 360
+        }
+        throw new Error('Should not be reachable.')
```
[the remark this answers is not in the document]
Comment (izulin) [reply src/interpreter/plugin/DatePlugin.ts:489]: mode is guaranteed to be one of {0,1,2,3,4} but TS doesn't understand that

[the remark this answers is not in the document]
Comment (izulin) [reply src/interpreter/plugin/DatePlugin.ts:489]: other solution here is to disable TS type system here, and also disable linter...

File: src/interpreter/plugin/DatePlugin.ts (hunk 1 above)
Comment (bardek8) [reply src/interpreter/plugin/DatePlugin.ts:94]: Good point! So we can mention that in the corresponding tests that we have:
`// Inconsistency with Product `
Do I understand correctly, that in such cases this would be safe to change (simply by removing the constraint on the argument) ?

File: src/interpreter/plugin/DatePlugin.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bardek8) [reply src/interpreter/plugin/DatePlugin.ts:489]: I thought that the user can pass e.g. 5 here and we should handle it appropriately

File: test/interpreter/function-days360.spec.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      ['=DAYS360("29/02/2020", "01/03/2020")'],
+      ['=DAYS360("28/02/2021", "01/03/2021")'],
+      ['=DAYS360("31/03/2020", "30/03/2020")'],
+      ['=DAYS360("29/02/2020", "28/02/2020")'],
+      ['=DAYS360("01/03/2020", "29/02/2020")'],
+      ['=DAYS360("01/03/2021", "28/02/2021")'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(0)
+    expect(engine.getCellValue(adr('A2'))).toEqual(3)
```
[the remark this answers is not in the document]
Comment (izulin) [reply test/interpreter/function-days360.spec.ts:40]: indeed, it is wrong

File: test/interpreter/function-days360.spec.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    expect(engine.getCellValue(adr('A3'))).toEqual(1)
+    expect(engine.getCellValue(adr('A4'))).toEqual(1)
+    expect(engine.getCellValue(adr('A5'))).toEqual(0)
+    expect(engine.getCellValue(adr('A6'))).toEqual(-2)
+    expect(engine.getCellValue(adr('A7'))).toEqual(0)
```
[the remark this answers is not in the document]
Comment (izulin) [reply test/interpreter/function-days360.spec.ts:45]: the same

File: test/interpreter/function-yearfrac.spec.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+      ['=YEARFRAC("29/02/2020", "01/03/2020")'],
+      ['=YEARFRAC("28/02/2021", "01/03/2021")'],
+      ['=YEARFRAC("31/03/2020", "30/03/2020")'],
+      ['=YEARFRAC("29/02/2020", "28/02/2020")'],
+      ['=YEARFRAC("01/03/2020", "29/02/2020")'],
+      ['=YEARFRAC("01/03/2021", "28/02/2021")'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(0)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(3/360, 9)
```
[the remark this answers is not in the document]
Comment (izulin) [reply test/interpreter/function-yearfrac.spec.ts:40]: indeed

File: test/interpreter/function-yearfrac.spec.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(1/366, 9)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(1/365, 9)
+    expect(engine.getCellValue(adr('A3'))).toBeCloseTo(2/366, 9)
+    expect(engine.getCellValue(adr('A4'))).toBeCloseTo(1/365, 9)
+    expect(engine.getCellValue(adr('A5'))).toBeCloseTo(61/366, 9)
+    expect(engine.getCellValue(adr('A6'))).toBeCloseTo(60/366, 9)
+    expect(engine.getCellValue(adr('A7'))).toBeCloseTo(59/365, 9)
+    expect(engine.getCellValue(adr('A8'))).toBeCloseTo(1, 9)
+    expect(engine.getCellValue(adr('A10'))).toBeCloseTo((366+365+365+365)/((366+365+365+365+366)/5), 9)
```
[the remark this answers is not in the document]
Comment (izulin) [reply test/interpreter/function-yearfrac.spec.ts:71]: yes

File: src/interpreter/plugin/DatePlugin.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (izulin) [reply src/interpreter/plugin/DatePlugin.ts:489]: the argument is defined as:

`argumentType: ArgumentTypes.INTEGER, defaultValue: 0, minValue: 0, maxValue: 4`

[the remark this answers is not in the document]
Comment (izulin) [reply src/interpreter/plugin/DatePlugin.ts:489]: (so any other value is handled earlier)

File: src/interpreter/plugin/DatePlugin.ts (hunk 1 above)
Comment (izulin) [reply src/interpreter/plugin/DatePlugin.ts:94]: yes

File: src/interpreter/plugin/DatePlugin.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bardek8) [reply src/interpreter/plugin/DatePlugin.ts:491]: Ok, you're right, this is checked there, I missed it.
However, this is the only non-zero boundary value set in the definition in the whole plugin and I didn't expect it there.

Actually it was the coverage test that pointed this line to me, not linter :)
